corehq/form_processor/models/forms.py

```diff
@@ -193,6 +193,25 @@ def get_deleted_form_ids_in_domain(self, domain):
             )
         return result
 
+    def hard_delete_forms_before_cutoff(self, cutoff, dry_run=True):
+        """
+        Permanently deletes forms with deleted_on set to a datetime earlier than
+        the specified cutoff datetime
+        :param cutoff: datetime used to obtain the forms to be hard deleted
+        :param dry_run: if True, no changes will be committed to the database
+        and this method is effectively read-only
+        :return: dictionary of count of deleted objects per table
+        """
+        counts = {}
+        for db_name in get_db_aliases_for_partitioned_query():
+            queryset = self.using(db_name).filter(deleted_on__lt=cutoff)
```

Comment:
Have you looked at the query plan for this on a large db e.g. prod?

`deleted_on` is not an indexed field so I think it will end up doing a full table scan which is going to be very slow & expensive.

We do have an index on the `state` column which also tracks deleted state so filtering by that should help, though the index is `domain, state` so you many need to filter by domain as well to get the index benefit.

Also, to filter by state you'll need to do a bit AND: `state & 64 = 64` since state field can represent multiple states e.g. archived & deleted.

You can see an example of that query using Django ORM here: https://github.com/dimagi/commcare-hq/blob/77652b164f6b396d3dbfbd7dd188842039dd00c1/corehq/form_processor/models/forms.py#L190-L191

Reply:
yes, we could do that

Reply:
Now that the partial index has been added we should verify the query plan for this to ensure that it uses the index. Assuming it does, I think this PR is good to go.

Reply:
The query as it current exists in the PR returns this:
```
                                                             QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using form_processor_xf_a66494ec_idx on form_processor_xforminstancesql  (cost=0.43..1231827.45 rows=1276719 width=664)
   Index Cond: (deleted_on < '2023-08-28 16:17:23.003229+00'::timestamp with time zone)
```